### ISSUE TITLE
crash in cleanDeviceMemoryManager() when multithreading

### ISSUE THREAD

**Issue by WilliamTambellini** [arrayfire/arrayfire#1814]
Hi all
AF devel sometimes crash in cleanDeviceMemoryManager() when multithreading.

```
OS: Ubuntu 16
backend: Cuda 8
ArrayFire v3.5.0 (CUDA, 64-bit Linux, build e59eb93)
Platform: CUDA Toolkit 8, Driver: 375.51
[0] GeForce GTX 1060, 6073 MB, CUDA Compute 6.1
```

Here is the bt of the crashing thread:
```
3  qsort  qsort.c                  79
8  cuMemFree_v2
9  cudart::cudaApiFree(void *)
10 cudaFree
11 cuda::MemoryManager::nativeFree(void *)
12 common::MemoryManager<cuda::MemoryManager>::cleanDeviceMemoryManager(int)
13 float * cuda::memAlloc<float>(unsigned long const&)
14 cuda::Array<float>::Array(af::dim4)
15 cuda::Array<float> cuda::createEmptyArray<float>(af::dim4 const&)
16 cuda::Array<float> cuda::matmul<float>(cuda::Array<float> const&, cuda::Array<float> const&, af_mat_prop, af_mat_prop) 
17 af_matmul
18 af_err unified::AFSymbolManager::call<void * *, void *, void *, af_mat_prop, af_mat_prop>(const char *, void * *, void *, void *, af_mat_prop, af_mat_prop)
19 af_matmul
20 af::matmul(af::array const&, af::array const&, af_mat_prop, af_mat_prop)
```

If that could help: the other threads were on the following tasks when the app crashed:
```
- cuda::initArray() > cuda::memAlloc() > lock_wait
- af::tile() > createBinaryNode > cuda::deviceMemoryInfo() > lock_wait
- af_retain_array > cuda::initArray > cuda::memAlloc() > lock wait
- af::sum > cuda::getKernel > cuda::compilekernel > cuLinkComplete > callJitEntryPoint
```

No I dont have steps to reproduce, this is reference for @umar456.

**Comment by pavanky** [comment]
@WilliamTambellini Are you sharing arrays across threads ? You can't give stack traces without background expecting a fix.

**Comment by WilliamTambellini** [comment]
Yep I m sharing some read only arrays across threads and AF was crashing whereas it was supposed to work.
I reproduced in front of @umar456  and we fixed it by adding some mutex locks in AF 3.5.
I suppose we could fix this one.
